### DISCUSSION TITLE
Middleware is worse than ORM hooks in the current state

### DISCUSSION THREAD

**Discussion by meotimdihia** [prisma/prisma#7992] [Q&A]
Prisma should be easier + faster in comparison with ORM in the theory
but middleware is worse than ORM hooks, it's very hard to write in middleware.
For example:
```javascript
const UserActionsMiddleware = async (params, next) => {
    // params.action could be "findUnique, findFirst, findMany, delete, deleteMany, update, updateMany, upsert" 
    // and params.args is different in the *many, upsert queries
```
while ORM hooks are simpler: `afterDelete`, `afterFind`, `afterUpdate`.

Another issue: https://github.com/prisma/prisma/issues/3398

**Comment by benjaminsuch** [comment]
I like the middleware-system the way it is right now, because of it's simplicity. I have a folder `db/middlewares` with all my middlewares and call them like:

```typescript
import { userActions, analysis } from 'db/middlewares'`

prisma.$use(userActions)

// or...

prisma.$use(async (params, next) => {
  const result = await next(params)
  return analysis(result)
})
```

For your `params.action` issue, I wrote a simple wrapper function where I just have to apply the actions I want my code to execute on. 

```typescript
prisma.$use(wrapper('findMany', userActions))
// or...
prisma.$use(wrapper(['findUnique', 'findFirst', 'findMany'], userActions))
prisma.$use(logger)
// or use it inside my middleware in case I want to execute different code based on the action...
const logger = async (params, next) => {
  const wrapperFn = wrapper(params, next)

  // The wrapper of course forwards `params` and `next` to the callback, but we can leave them blank here,
  // since we have them already provided by `logger`.
  wrapperFn('delete', async (/*params, next*/) => {
    console.log(`Deleting item from table with id ${params.args.id}`)
  })

  wrapperFn(['findMany', 'findUnique', 'findFirst'], async () => {
    // ...
  })
}

const userActions = async (params, next) => {
  // ...
}
```

Sure you could argue, that this would bloat up the middleware, but I think this depends how you structure and write code.

**Reply by meotimdihia** [reply]
@benjaminsuch: Could you show me how to handle my case in my reply to @janpio?

**Comment by janpio** [comment]
Could you maybe write a middleware that creates and sends these hooks you are longing for? 

We are aware that middlewares are a "build it yourself" solution taht we chose to give people all the power to come up with their own way of doing things. If hooks are much better for some use cases, we are certainly open to consider them as well. We would need a feature request issue that spells out the problem to solve and makes a few suggestion on how hooks could look like then (we might even already have one).

**Reply by meotimdihia** [reply]
Currently, Middleware really doesn't work.  For example my query: 

```js
      await prisma.chapters.update({
          data: {
            chap: chapterData.attributes.chapter,
            images:
              images.length > 0
                ? {
                    deleteMany: {
                      OR: [
                        {
                          AND: [
                            {
                              name: {
                                notIn: images,
                              },
                            },
                          ],
                        },
                        ...
```

Won't trigger `deleteMany` of ImagesMiddleware: 

```js
const ImagesMiddleware = async (params, next) => {
  if (
    params.model == "images" &&
    ["delete", "deleteMany"].indexOf(params.action) != -1
```

There is no way to catch all cases for this.

**Comment by ludralph** [comment]
Hi there,

To keep our discussions organized and focused on the most relevant topics, we’re reviewing and tidying up our backlog. As part of this process, we’re closing discussions that haven’t had any recent activity and appear to be outdated.

If this discussion is still important to you or unresolved, we’d love to hear from you! Feel free to reopen it or start a new one with updated details.

For more details about our priorities and vision for the future of Prisma ORM, check out our latest blog post: https://www.prisma.io/blog/prisma-orm-manifesto.

Thank you for your understanding and being part of the community!
